### ISSUE TITLE
Add support for pass-otp plugin

### ISSUE THREAD

**Issue by balasankarc** [passff/passff#200]
It would be amazing if we could add support for [pass-otp plugin](https://github.com/tadfisher/pass-otp) also, which provides OTP ability.

**Comment by nwallace** [comment]
Interesting. Hadn't heard of this plugin before. I'm not all that familiar with how OTPs work, and I'm not getting it from the project's README. Can you ELI5 how OTPs work and what PassFF might do with this pass plugin?

**Comment by balasankarc** [comment]
Disclaimer: I have no idea how PassFF works internally. 

OTPs (One Time Passwords) are a string generated based on an `otpauth` URL and the time of generation. These are used in 2 factor authentications. The generated OTPs work for a certain period of time only, hence providing security.

pass-otp is a CLI tool that uses `oathtool` to generate the OTPs. If a password file has an `otpauth` URL defind, `pass otp <password file>` will get you an OTP code.

(If PassFF shells out to `pass` command in the background - which I don't assume it is - the whole process would be easier. You just have to shell out `pass top <file>` and get the returned value. But if PassFF actually parses the password file, without using the `pass` tool, it will be difficult. You will have to incorporate an OTP generation logic)

The above are my basic understanding of OTP. Internet searches will be much more helpful than me. I will try to get you some links when I get some free time.

**Comment by tuxor1337** [comment]
Actually, PassFF does shell out (contrary to other pass-based browser plugins) although the specific commands are hardcoded to prevent code injection. So, as you say, the implementation of the OTP plugin into PassFF would be mostly a matter of incorporating it into the UI. Any suggestions what this could look like?

**Comment by phryneas** [comment]
Essentially, as most modern OTP is following the time-based TOTP standard, ideally it would display a OTP-Code (usually a 6-digit number) as well as the time it was generated.

Or just enter the code into an input field - it would be another input field like "login" or "password"

**Comment by tuxor1337** [comment]
I don't think you got me right: I'm asking how exactly the user chooses to use the OTP plugin. How and when does she tell PassFF to use the OTP plugin?

**Comment by phryneas** [comment]
Yup, realized that a moment later.

Usually, a user will either enter a password as well as an OTP token. 
Either into the same form, or into pages following each other.

If I interpret the code of the plugin right, both the password and the othauth-code can be stored in the same password file, as there is no need for the otpauth url to be in the first line.

So there might be three useful entries in the UI:

* fill in password (no OTP necessary)
* fill in OTP-token (always OTP)
* fill in everything (invoke OTP if any line in the password file begins with  `otpauth://totp/totp-secret?`  )

**Comment by nwallace** [comment]
Ok, I think I get it. So pass-otp can generate the token and put it into your password file. Then if we could read that token out of the file for copy/pasting, auto-filling, etc, it would be easier to use pass-otp with passff?

**Comment by phryneas** [comment]
The token is created from the secret stored in the line beginning with `otpauth://totp/totp-secret?` the moment when pass-otp is called. Calling pass-otp in this fashion may have side-effects (increase a counter on client side for HOTP), so pass-otp for HOTP should only be called when explicitly requested by the user and not by accident. TOTP could be handled a bit more loosely.

Just as a short explanation - there are two types of those tokens:

TOTP: a token is calculated as a HMAC of `a secret shared by client and server` and `the current time`
HOTP: a token is calculated as a HMAC of `a secret shared by client and server` and `a counter that is incremented on client & server`

As server and client time may differ, usually the time is rounded to the next five minutes and the server verifies against the currently correct TOTP token as well as the last two and the next two.
Same with HOTP. Server and client may get out of sync, so usually a few tokens in the future are accepted as well, and the server corrects the counter.

When you use second factor for facebook, paypal or github, that is usually a TOTP token. HOTP tokens are quite rare nowadays.

**Comment by balasankarc** [comment]
Meanwhile lemme state that I'm extremely happy that there's a discussion around it already. This is usually not the case with other FOSS tools I engaged with, where I should at least bump the issue twice to get a response.

You guys rock. 👍

**Comment by tadfisher** [comment]
If it helps, external clients (i.e. passff) should never have to deal with the OTP secret directly; `pass otp show $passfile` should be all that is required for integration on the client side. This writes the generated OTP code to stdout.

Adding new OTP secrets would be nice, and is provided for in the `pass-otp` interface by piping an `otpauth://` URI into `pass otp insert` or `pass otp append`. Browser integration would make this very convenient as you could perform QR decoding, for example.

**Comment by phryneas** [comment]
The thing is that passff has to somehow discover if an entry can be used with `pass otp` - I guess decrypting it and looking for the otpauth:// url is the only way for that, isn't it?

**Comment by tadfisher** [comment]
Yep, that was a conscious decision. If it helps, `pass-otp` only respects lines that consist solely of an `otpauth://` url; no prefix or suffix.

**Comment by balasankarc** [comment]
@phryneas Check if `pass otp <filename>` returns an empty string? We need not decrypt the file.

**Comment by phryneas** [comment]
@balasankarc if the OTP token in that file is a HOTP token, this simple check will increase the counter by 1 and after a few such checks, server and client go out of sync and the `pass otp` generated tokens won't work any more.

Maybe just a button that either returns a token or an error message, without prior checking, might be the best idea.

**Comment by balasankarc** [comment]
@phryneas Oh. I missed `HOTP`. Sorry. :)

**Comment by tadfisher** [comment]
If there's anything I can do to make this easier, such as a command to list
all passfiles with OTP, or one to check for the existence of an OTP URI in
a single passfile, I'd be happy to do the work. Let me know what works best
for passff.

**Comment by tadfisher** [comment]
> The thing is, can you please implement it in a way that does not require looking into every single file? Something like keeping a central (encrypted) file of otp-able entries. Otherwise I fear that it could take quite a while to scan through a big password store (I have almost 600 entries)

Agreed, I'll look into this.

**Comment by tadfisher** [comment]
@phryneas Can you run `pass grep "some string"` and let me know how long that takes? That would be a baseline measurement of the naive "scan file-by-file" approach.

**Comment by phryneas** [comment]
Well, I'm kinda the worst-case scenario, as I'm using crypto-hardware (smartcard, yubikey) on windows systems (slower forking): 

* using a GnuPG smartcard on my laptop on windows: 4:33
* using a Yubikey on my PC on windows: made multiple attempts, but after two minutes of steadily decrypting files, the yubikey connection is lost and even re-inserting does not allow me to continue.

**Comment by tadfisher** [comment]
Timed using 4096-bit keys on a Yubikey 4: `pass grep` on 30 entries took 30 seconds. Agreed, this is something I'll work on.

**Comment by nwallace** [comment]
Correct me if I'm wrong (I may not understand everything right), but I think we won't need to decrypt the whole password store. I think we just need to know if the user has `pass-otp`. We could do this via automatic detection (via `pass otp --help` or something) or with a preference in the plugin.  Once we know that the user has `pass-otp`, we can add stuff like a "Copy OTP" option. Even if the action "Copy OTP" is displayed for all your passwords (including passwords that aren't using OTP), you would be smart enough to know when to use that option and when not to.

**Comment by tadfisher** [comment]
@nwallace That's certainly an option. I actually started working on shell completion support for extensions, and a listing of extensions in `pass help` would be a nice option for you to use. This is how completion for `git` extensions work, anyhow.

**Comment by SpEcHiDe** [comment]
I happened to come across this site: https://github.com/Daplie/browser-authenticator and I think it would help you. The OTP in the above link can be generated using pure JavaScript.

**Comment by balasankarc** [comment]
Thanks @SpEcHiDe for that info. But I would personally prefer the `pass` plugin so that I need not use multiple tools. :)

**Comment by SpEcHiDe** [comment]
What I meant was the `browserpass` installed on a system can interact with the `pass` utility and get the line saved in the password file. After getting a valid otpauth:// URI, the token can be generated using the above JavaScript. Hope I am clear. Since, I do not have any experience developing FireFox extensions, I decided to put the link here, for the developers to see.

**Comment by cjustin88** [comment]
I also vouch for this feature request.

**Comment by grimpy** [comment]
https://gist.github.com/grimpy/8239665dfd7097a72543e2407747a5f9

Quick hack, basicly copies totp instead of password if file contains totp

**Comment by tuxor1337** [comment]
You mean: `if file name contains totp`. It's not a solution for this issue, but - as you said - a quick hack that might be interesting for some users. Thanks!

**Comment by madduck** [comment]
For your information, https://github.com/dannyvankooten/browserpass solved this by displaying a little overlay with the current OTP in the corner, see a screenshot here: https://github.com/dannyvankooten/browserpass/issues/187#issuecomment-342952258

This requires some manual interaction, but it's better than using the terminal right now. And maybe we can figure out how to auto-fill after all, i.e. using custom field names…

```
mypasswd
login: madduck
otpauth://…/…
otpfield: #googleauth
```

**Comment by LucidOne** [comment]
:rotating_light: The pull request needs volunteers for testing. :rotating_light: 

Bumping this issue as I'm not sure if Github sends notifications to non-maintainers when an issue is referenced in a pull request.
